src/common/hashtable.c

```diff
@@ -0,0 +1,351 @@
+    const int new_size = ht->base_size * 2;
+    return onf_ht_resize(ht, new_size);
+}
+
+/*****************************************
+Name:  onf_ht_resize_down
+brief: resize down
+return: 
+*****************************************/
+int onf_ht_resize_down(onf_ht_hash_table* ht) {
+    const int new_size = ht->base_size / 2;
+    return onf_ht_resize(ht, new_size);
+}
+
+/*****************************************
+Name:  onf_ht_del_item
+brief: delete hash item
+return: 
+*****************************************/
+void onf_ht_del_item(onf_ht_item* i) {
+    free(i->key);
+    free(i->value);
+    free(i);
+}
+
+/*****************************************
+Name:  onf_ht_del_hash_table
+brief: delete hash table
+return: 
+*****************************************/
+void onf_ht_del_hash_table(onf_ht_hash_table* ht) {
+    for (int i = 0; i < ht->size; i++) {
+        onf_ht_item* item = ht->items[i];
+        if (item != NULL) {
+            onf_ht_del_item(item);
+        }
+    }
+    free(ht->items);
+    free(ht);
+}
+
+/*****************************************
+Name:  onf_ht_hash
+brief: delete hash table
+return: 
+*****************************************/
+int onf_ht_hash(
+      const char* s, int len_s, const int a, const int m) 
+{
+    long hash = 0;
+    for (int i = 0; i < len_s; i++) {
+        hash += (long)pow(a, len_s - (i+1)) * s[i];
+        hash = hash % m;
+    }
+    return (int)hash;
+}
+
+/*****************************************
+Name:  onf_ht_get_hash
+brief: get hash table
+return: 
+*****************************************/
+int onf_ht_get_hash(
+    const char* s, int key_len, 
+    const int num_buckets, const int attempt) 
+{
+    const int hash_a = onf_ht_hash(
+                        s, key_len, ONF_HT_PRIME_1, num_buckets);
+    const int hash_b = onf_ht_hash(
+                        s, key_len, ONF_HT_PRIME_2, num_buckets);
+    return (hash_a + (attempt * (hash_b + 1))) % num_buckets;
+}
+
+/*****************************************
+Name:  onf_ht_insert
+brief: insert into hash table
+return: 
+*****************************************/
+int onf_ht_insert(
+      onf_ht_hash_table* ht, int key_len,  
+      int val_len, const char* key, const char* value) 
+{
+    const int load = (ht->count * 100) / ht->size;
+    if (load > 70) {
+        if(SUCCESS != onf_ht_resize_up(ht))
+	{
+	    return E_FAIL;
+	}
+    }
+    
+    onf_ht_item* item = onf_ht_new_item(
+                         key_len, val_len,key, value);
+    if(NULL == item)
+    {
+       return E_ALLOC_FAILED;
+    }
+    int index = onf_ht_get_hash(item->key, 
+                   item->key_len,ht->size, 0);
+    onf_ht_item* cur_item = ht->items[index];
+    int i = 1;
+    
+    while (cur_item != NULL) {
+        if(cur_item != &ONF_HT_DELETED_ITEM) {
+            if (onf_key_compare(
+	          cur_item->key_len, key_len,
+		      cur_item->key, key)) {
+                onf_ht_del_item(cur_item);
```

Comment:
Is this duplicate element ?

Reply:
Yes, Its for an update case. To change the value for the existing key.